rbd/watchers_mimic.go

```diff
@@ -0,0 +1,98 @@
+// +build !luminous,!mimic
+//
+// Ceph Nautilus is the first version that supports watchers through librbd.
+
+package rbd
+
+// #cgo LDFLAGS: -lrbd
+// #include <errno.h>
+// #include <rbd/librbd.h>
+import "C"
+
+// Handle for a watcher
+// This type was added so that Image.HasWatchers() can be tested.
+type rbdUpdateHandle struct {
+	handle C.uint64_t
+}
+
+// TODO: rename to UpdateWatch() and accept callback arguments
+// This function was added so that Image.HasWatchers() can be tested.
+//
+// Implements:
+//   int rbd_update_watch(rbd_image_t image, uint64_t *handle,
+//                        rbd_update_callback_t watch_cb, void *arg)
+func (image *Image) updateWatch() (*rbdUpdateHandle, error) {
+	if err := image.validate(imageIsOpen); err != nil {
+		return nil, err
+	}
+
+	ruh := &rbdUpdateHandle{}
+
+	// TODO: passing callbacks from Go to C is tricky, see
+	// https://github.com/golang/go/wiki/cgo#function-pointer-callbacks
+	ret := C.rbd_update_watch(image.image, &ruh.handle, nil, nil)
+	if ret != 0 {
+		return nil, getError(ret)
+	}
+
+	return ruh, nil
+}
+
+// TODO: rename to UpdateUnwatch()
+// This function was added so that Image.HasWatchers() can be tested.
+// FIXME: this sometimes panics, use Image.Close() instead (for now)
+//
+// Implements:
+//   int rbd_update_unwatch(rbd_image_t image, uint64_t handle)
+func (image *Image) updateUnwatch(ruh *rbdUpdateHandle) error {
+	if err := image.validate(imageIsOpen); err != nil {
+		return err
+	}
+
+	ret := C.rbd_update_unwatch(image.image, ruh.handle)
+	if ret != 0 {
+		return getError(ret)
+	}
+
+	return nil
+}
+
+// ListWatchers returns the watchers on an RBD image. In case of an error, nil
+// and an error are returned.
+//
+// Note:
+//   Only supported in Ceph Nautilus and newer.
+//
+// Implements:
+//   int rbd_watchers_list(rbd_image_t image,
+//                         rbd_image_watcher_t *watchers, size_t *max_watchers)
+func (image *Image) ListWatchers() ([]ImageWatcher, error) {
+	if err := image.validate(imageIsOpen); err != nil {
+		return nil, err
+	}
+
+	count := C.ulong(0)
+	ret := C.rbd_watchers_list(image.image, nil, &count)
```

Comment:
It just occurred to me that there's may be a chance of "race condition" between the call that fetches the size/count of watchers vs. actually fetching the list. New watchers might be added between the calls and then then the 2nd call will fail (that's OK but suboptimal IMO). There's a couple of things we might want to do do mitigate that, but I won't block this PR over this. However, if you choose not to take the task on in this PR please let's file a new issue to track it.

Reply:
yes, these kind of races are present in other parts of the code as well. There is not much to do about it, as the watchers can change after this function returns and before the application acts on it. I'd prefer to see if we can use a common way of doing the ERANGE check and replace the occurrences all together.

Reported as #185

Reply:
OK, some of the function calls vary in how they work with ERANGE so it may not be easy to share code, but we can certainly think about how to approach it in a common enough way.